lib/web_ui/lib/src/engine/html/picture.dart

```diff
@@ -603,3 +630,44 @@ class PersistedPicture extends PersistedLeafSurface {
     }
   }
 }
+
+/// Given size of a rectangle and transform, computes pixel density
+/// (scale factor).
+double _computePixelDensity(Matrix4? transform, double width, double height) {
+  if (transform == null || transform.isIdentity()) {
+    return 1.0;
+  }
+  final Float32List m = transform.storage;
+  // Apply perspective transform.
+  final double topLeftX = m[12] * m[15];
+  final double topLeftY = m[13] * m[15];
+  double x = width;
+  double y = height;
+  final double xp = (m[0] * x) + (m[4] * y) + m[12];
+  final double yp = (m[1] * x) + (m[5] * y) + m[13];
+  final double wp = 1.0 / ((m[3] * x) + (m[7] * y) + m[15]);
+  double scaleX = ((xp * wp) - topLeftX).abs() / width;
+  double scaleY = ((yp * wp) - topLeftY).abs() / height;
+  double scale = math.min(scaleX, scaleY);
+  if (scale < kEpsilon || scale == 1) {
+    // Handle local paint bounds scaled to 0, typical when using
+    // transform animations.
+    return 1.0;
```

Comment:
Shouldn't this be `if ((scale - 1).abs() < kEpsilon)`? Also scale 0 probably means "empty picture".

Reply:
Added and updated comment. It guards the code below from divide by zero.